### ISSUE TITLE
Need to describe type for cypress module API

### ISSUE THREAD

**Issue by bahmutov** [cypress-io/cypress#2141]
When using Cypress via its module API, the resolved value is quite large and complex. So we should include typescript definitions for it with the code. This will be especially useful for the output of the `cypress.run` method.

Example `cypress.run` resolved value

```json
{
  "startedTestsAt": "2018-07-13T05:36:53.230Z",
  "endedTestsAt": "2018-07-13T05:37:14.954Z",
  "totalDuration": 21724,
  "totalSuites": 0,
  "totalTests": 1,
  "totalFailed": 0,
  "totalPassed": 1,
  "totalPending": 0,
  "totalSkipped": 0,
  "runs": [
    {
      "stats": {
        "suites": 0,
        "tests": 1,
        "passes": 1,
        "pending": 0,
        "skipped": 0,
        "failures": 0,
        "wallClockStartedAt": "2018-07-13T05:36:53.230Z",
        "wallClockEndedAt": "2018-07-13T05:37:14.954Z",
        "wallClockDuration": 21724
      },
      "reporter": "spec",
      "reporterStats": {
        "suites": 0,
        "tests": 1,
        "passes": 1,
        "pending": 0,
        "failures": 0,
        "start": "2018-07-13T05:36:53.238Z",
        "end": "2018-07-13T05:37:14.983Z",
        "duration": 21745
      },
      "hooks": [],
      "tests": [
        {
          "testId": "r2",
          "title": [
            "adds and removes completed todos"
          ],
          "state": "passed",
          "body": "function () {\n  var _this = this;\n\n  cy.visit('http://todomvc.com/examples/react/', {\n    onBeforeLoad: function onBeforeLoad(win) {\n      loadTinyToast(win.document);\n    },\n    onLoad: function onLoad(win) {\n      win.tinyToast.show(_this.test.title).hide(3000);\n    }\n  });\n  cy.get('.new-todo').should('be.visible');\n  addTodo('this is first todo');\n  addTodo('second todo');\n\n  // cy.wait(1000)\n  cy.contains('ul.todo-list li', 'second todo').find('input.toggle').check();\n  cy.toast('Checked 1 todo');\n  scroll();\n\n  cy.contains('.todo-count', '1').should('be.visible');\n\n  cy.toast('Filters: Active / Completed / All');\n  pickFilter('Active');\n  pickFilter('Completed');\n  pickFilter('All');\n\n  cy.toast('Clicking \"Clear completed\" button');\n  clearCompleted();\n\n  cy.toast('Completed items have been removed', 5000);\n\n  cy.then(function () {\n    var rect = findAppRectangle();\n    var size = serializeRectangle(rect);\n    console.log(size);\n    // return cy.task('size', size)\n    var s = JSON.stringify(size, null, 2) + '\\n';\n    cy.writeFile('size.json', s, 'utf8');\n  });\n\n  // capture several resolutions\n  // cy.screenshot('runner', {\n  //   capture: 'runner'\n  // })\n  // cy.screenshot('app', {\n  //   capture: 'viewport'\n  // })\n  // cy.screenshot('full', {\n  //   capture: 'fullPage'\n  // })\n}",
          "stack": null,
          "error": null,
          "timings": {
            "lifecycle": 28,
            "test": {
              "fnDuration": 20067,
              "afterFnDuration": 1
            }
          },
          "failedFromHookId": null,
          "wallClockStartedAt": "2018-07-13T05:36:54.852Z",
          "wallClockDuration": 20098,
          "videoTimestamp": 3770
        }
      ],
      "error": null,
      "video": "/Users/gleb/git/cut-movie/cypress/videos/spec.js.mp4",
      "screenshots": [],
      "spec": {
        "name": "spec.js",
        "relative": "cypress/integration/spec.js",
        "absolute": "/Users/gleb/git/cut-movie/cypress/integration/spec.js"
      },
      "shouldUploadVideo": true
    }
  ],
  "browserPath": "",
  "browserName": "electron",
  "browserVersion": "59.0.3071.115",
  "osName": "darwin",
  "osVersion": "16.7.0",
  "cypressVersion": "3.0.2",
  "config": {
    "viewportWidth": 1000,
    "viewportHeight": 1000,
    "videoCompression": false,
    "projectRoot": "/Users/gleb/git/cut-movie",
    "projectName": "cut-movie",
    "morgan": false,
    "isTextTerminal": true,
    "socketId": "q34tb",
    "report": true,
    "browsers": [
      {
        "name": "chrome",
        "displayName": "Chrome",
        "version": "67.0.3396.99",
        "path": "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
        "majorVersion": "67"
      },
      {
        "name": "electron",
        "displayName": "Electron",
        "version": "59.0.3071.115",
        "path": "",
        "majorVersion": "59",
        "info": "Electron is the default browser that comes with Cypress. This is the browser that runs in headless mode. Selecting this browser is useful when debugging. The version number indicates the underlying Chromium version that Electron uses."
      }
    ],
    "port": 56514,
    "hosts": null,
    "baseUrl": null,
    "userAgent": null,
    "reporter": "spec",
    "reporterOptions": null,
    "blacklistHosts": null,
    "clientRoute": "/__/",
    "xhrRoute": "/xhrs/",
    "socketIoRoute": "/__socket.io",
    "socketIoCookie": "__socket.io",
    "reporterRoute": "/__cypress/reporter",
    "ignoreTestFiles": "*.hot-update.js",
    "testFiles": "**/*.*",
    "defaultCommandTimeout": 4000,
    "requestTimeout": 5000,
    "responseTimeout": 30000,
    "pageLoadTimeout": 60000,
    "execTimeout": 60000,
    "taskTimeout": 60000,
    "video": true,
    "videoUploadOnPasses": true,
    "modifyObstructiveCode": true,
    "chromeWebSecurity": true,
    "waitForAnimations": true,
    "animationDistanceThreshold": 5,
    "numTestsKeptInMemory": 0,
    "watchForFileChanges": false,
    "trashAssetsBeforeRuns": true,
    "autoOpen": false,
    "fileServerFolder": "/Users/gleb/git/cut-movie",
    "videosFolder": "/Users/gleb/git/cut-movie/cypress/videos",
    "supportFile": "/Users/gleb/git/cut-movie/cypress/support/index.js",
    "fixturesFolder": "/Users/gleb/git/cut-movie/cypress/fixtures",
    "integrationFolder": "/Users/gleb/git/cut-movie/cypress/integration",
    "screenshotsFolder": "/Users/gleb/git/cut-movie/cypress/screenshots",
    "namespace": "__cypress",
    "pluginsFile": "/Users/gleb/git/cut-movie/cypress/plugins/index.js",
    "javascripts": [],
    "env": {},
    "cypressEnv": "production",
    "resolved": {
      "animationDistanceThreshold": {
        "value": 5,
        "from": "default"
      },
      "fileServerFolder": {
        "value": "",
        "from": "default"
      },
      "baseUrl": {
        "value": null,
        "from": "default"
      },
      "fixturesFolder": {
        "value": "cypress/fixtures",
        "from": "default"
      },
      "chromeWebSecurity": {
        "value": true,
        "from": "default"
      },
      "modifyObstructiveCode": {
        "value": true,
        "from": "default"
      },
      "integrationFolder": {
        "value": "cypress/integration",
        "from": "default"
      },
      "env": {},
      "pluginsFile": {
        "value": "cypress/plugins",
        "from": "default"
      },
      "hosts": {
        "value": null,
        "from": "default"
      },
      "screenshotsFolder": {
        "value": "cypress/screenshots",
        "from": "default"
      },
      "numTestsKeptInMemory": {
        "value": 0,
        "from": "config"
      },
      "supportFile": {
        "value": "cypress/support",
        "from": "default"
      },
      "port": {
        "value": null,
        "from": "default"
      },
      "reporter": {
        "value": "spec",
        "from": "default"
      },
      "videosFolder": {
        "value": "cypress/videos",
        "from": "default"
      },
      "reporterOptions": {
        "value": null,
        "from": "default"
      },
      "testFiles": {
        "value": "**/*.*",
        "from": "default"
      },
      "defaultCommandTimeout": {
        "value": 4000,
        "from": "default"
      },
      "trashAssetsBeforeRuns": {
        "value": true,
        "from": "default"
      },
      "execTimeout": {
        "value": 60000,
        "from": "default"
      },
      "blacklistHosts": {
        "value": null,
        "from": "default"
      },
      "pageLoadTimeout": {
        "value": 60000,
        "from": "default"
      },
      "userAgent": {
        "value": null,
        "from": "default"
      },
      "requestTimeout": {
        "value": 5000,
        "from": "default"
      },
      "viewportWidth": {
        "value": 1000,
        "from": "default"
      },
      "responseTimeout": {
        "value": 30000,
        "from": "default"
      },
      "viewportHeight": {
        "value": 1000,
        "from": "config"
      },
      "taskTimeout": {
        "value": 60000,
        "from": "default"
      },
      "video": {
        "value": true,
        "from": "default"
      },
      "videoCompression": {
        "value": false,
        "from": "config"
      },
      "videoUploadOnPasses": {
        "value": true,
        "from": "default"
      },
      "watchForFileChanges": {
        "value": false,
        "from": "config"
      },
      "waitForAnimations": {
        "value": true,
        "from": "default"
      }
    },
    "parentTestsFolder": "/Users/gleb/git/cut-movie/cypress",
    "parentTestsFolderDisplay": "cut-movie/cypress",
    "supportFolder": "/Users/gleb/git/cut-movie/cypress/support",
    "integrationExampleName": "examples",
    "integrationExamplePath": "/Users/gleb/git/cut-movie/cypress/integration/examples",
    "scaffoldedFiles": [
      {
        "name": "cypress",
        "children": [
          {
            "name": "integration",
            "children": [
              {
                "name": "examples",
                "children": [
                  {
                    "name": "actions.spec.js"
                  },
                  {
                    "name": "aliasing.spec.js"
                  },
                  {
                    "name": "assertions.spec.js"
                  },
                  {
                    "name": "connectors.spec.js"
                  },
                  {
                    "name": "cookies.spec.js"
                  },
                  {
                    "name": "cypress_api.spec.js"
                  },
                  {
                    "name": "files.spec.js"
                  },
                  {
                    "name": "local_storage.spec.js"
                  },
                  {
                    "name": "location.spec.js"
                  },
                  {
                    "name": "misc.spec.js"
                  },
                  {
                    "name": "navigation.spec.js"
                  },
                  {
                    "name": "network_requests.spec.js"
                  },
                  {
                    "name": "querying.spec.js"
                  },
                  {
                    "name": "spies_stubs_clocks.spec.js"
                  },
                  {
                    "name": "traversal.spec.js"
                  },
                  {
                    "name": "utilities.spec.js"
                  },
                  {
                    "name": "viewport.spec.js"
                  },
                  {
                    "name": "waiting.spec.js"
                  },
                  {
                    "name": "window.spec.js"
                  }
                ]
              }
            ]
          },
          {
            "name": "fixtures",
            "children": [
              {
                "name": "example.json"
              }
            ]
          },
          {
            "name": "support",
            "children": [
              {
                "name": "commands.js"
              },
              {
                "name": "index.js"
              }
            ]
          },
          {
            "name": "plugins",
            "children": [
              {
                "name": "index.js"
              }
            ]
          }
        ]
      }
    ],
    "state": {},
    "isNewProject": false,
    "proxyUrl": "http://localhost:56514",
    "browserUrl": "http://localhost:56514/__/",
    "reporterUrl": "http://localhost:56514/__cypress/reporter",
    "xhrUrl": "__cypress/xhrs/"
  }
}
```

**Comment by jennifer-shehane** [comment]
Yes, I actually had to update our docs recently for the returned value, as it had changed in 3.0 and it was not updated, so +1 for TypeScript - this may have helped this user catch it quicker. https://github.com/cypress-io/cypress/issues/2125

**Comment by cdaringe** [comment]
i ran into this today.  in fact, you can't import the module api in typescript because no module is exported from the typings either.   when this happens, the module type should be exported from the typings as well!  i'll test it :)

**Comment by jennifer-shehane** [comment]
Released in [3.3.0](https://on.cypress.io/changelog#3-0-0)
